File: 02/solution.exs
defmodule Solution do
  def part_one(path) do
    solution(path, &safe_part_one?/1)
  end

  def part_two(path) do
    solution(path, &safe_part_two?/1)
  end

  defp solution(path, safe?) do
    path
    |> File.stream!()
    |> Stream.map(&read_levels/1)
    |> Stream.filter(safe?)
    |> Enum.count()
    |> IO.puts()
  end

  defp read_levels(line) do
    line
    |> String.trim("\n")
    |> String.split(" ", trim: true)
    |> Enum.map(&String.to_integer/1)
  end

  defp safe_part_one?(level) do
    moderately_increasing?(level) or moderately_increasing?(Enum.reverse(level))
  end

  defp safe_part_two?(level) do
    if safe_part_one?(level) do
      true
    else
      0..(length(level) - 1)
      |> Stream.map(&without_index(level, &1))
      |> Enum.any?(&safe_part_one?/1)
    end
  end

  defp moderately_increasing?(level) do
    levels_firsts = Enum.slice(level, 0..-2//1)
    levels_lasts = Enum.slice(level, 1..-1//1)

    [levels_firsts, levels_lasts]
    |> List.zip()
    |> Enum.all?(fn {l1, l2} -> l2 > l1 and l2 <= l1 + 3 end)
  end

  defp without_index(level, i) do
    level
    |> Enum.with_index()
    |> Enum.filter(fn {_, j} -> i != j end)
    |> Enum.map(fn {elem, _} -> elem end)
  end
end

path = System.argv() |> Enum.at(0)
Solution.part_one(path)
Solution.part_two(path)

File: 02/solution.py
from sys import argv


def part_one(path):
    solution(path, is_safe)


def part_two(path):
    solution(path, is_safe_with_tolerance)


def solution(path, is_safe_func):
    print(sum(is_safe_func(levels) for levels in read_levels(path)))


def read_levels(path):
    with open(path, "r") as f:
        for line in f:
            yield [int(l) for l in line.strip().split()]


def is_safe(levels):
    return is_moderately_increasing(levels) or is_moderately_increasing(levels[::-1])


def is_safe_with_tolerance(levels):
    if is_safe(levels):
        return True
    return any(is_safe(truncated_level) for truncated_level in truncate_levels(levels))


def is_moderately_increasing(levels):
    growth_rates = (
        next_level - current_level
        for (current_level, next_level) in zip(levels[:-1], levels[1:])
    )
    res = all(1 <= growth_rate <= 3 for growth_rate in growth_rates)
    return res


def truncate_levels(levels):
    for i, _ in enumerate(levels):
        yield levels[:i] + levels[i + 1 :]


if __name__ == "__main__":
    path = argv[1]
    part_one(path)
    part_two(path)
